dockers/docker-sonic-telemetry/certificate_rotation_checker

```diff
@@ -0,0 +1,225 @@
+    """
+    processes_status_cmd = "supervisorctl status"
+    retry_times = 0
+    is_running = False
+
+    while retry_times <= MAX_RETRY_TIMES:
+        retry_times += 1
+        exit_code, command_stdout = get_command_result(processes_status_cmd)
+        if exit_code != 3:
+            syslog.syslog(syslog.LOG_INFO,
+                          "Failed to get the processes running status in telemetry container and retry after 60 seconds ...")
+            time.sleep(60)
+        else:
+            for line in command_stdout.splitlines():
+                if "telemetry" in line and "RUNNING" in line:
+                    is_running = True
+                    break
+            if is_running:
+                syslog.syslog(syslog.LOG_INFO,
+                              "Telemetry server process is running after certificate and private key were rotated!")
+                break
+
+    if not is_running:
+        syslog.syslog(syslog.LOG_ERR,
+                      "Telemetry server process is not running after certificate and private key were rotated and exiting ...")
+        sys.exit(1)
+
+
+def restart_telemetry_server():
+    """Restarts the telemetry server process by Supervisord and then checks
+    it is actually running.
+
+    Args:
+        None
+
+    Returns:
+        None
+    """
+    restart_telemetry_server_cmd = "supervisorctl restart telemetry"
+    retry_times = 0
+
+    while retry_times <= MAX_RETRY_TIMES:
+        retry_times += 1
+        exit_code, command_stdout = get_command_result(restart_telemetry_server_cmd)
+        if exit_code != 0:
+            syslog.syslog(syslog.LOG_INFO,
+                          "Failed to restart telemetry server process and retry after 60 seconds ...")
+            time.sleep(60)
+        else:
+            break
+
+    if retry_times > MAX_RETRY_TIMES:
+        syslog.syslog(syslog.LOG_ERR,
+                      "Failed to restart telemetry server process after trying '{}' times and exiting ..."
+                      .format(MAX_RETRY_TIMES))
+        sys.exit(2)
+
+    check_telemetry_server_running()
+
+
+def check_certificate_rotated(certificate_file_name, private_key_file_name):
+    """Leverages the 'inotify' module to monitor the file system events under the
+    directory which stores the SONiC credentials and restarts telemetry server
+    process if certificate and private key were rotated.
+
+
+    Args:
+        certificate_file_name: A string indicates the telemetry certificate file name.
+        private_key_file_name: A string indicates the telemetry private key file name.
+
+    Returns:
+        None.
+    """
+    certificate_file_rotated = False
+    private_key_file_rotated = False
+
+    inotify_instance = inotify.adapters.Inotify()
+    inotify_instance.add_watch(CREDENTIALS_DIR_PATH)
+    for event in inotify_instance.event_gen(yield_nones=False):
+        header, event_type, monitoring_path, file_name = event
+        if (file_name == certificate_file_name
```

Comment:
>if [](http://example.com/codeflow?start=8&length=2)

The FSM logic is complex and may be messed up by some input sequence. Could you use one file as the main indicator, and always rotate if that file changed.

ref: [https://github.com/Azure/sonic-restapi/blob/94805a39ac0712219f7dc08faa2cfdbf371dd177/go-server-server/main.go#L103](https://github.com/Azure/sonic-restapi/blob/94805a39ac0712219f7dc08faa2cfdbf371dd177/go-server-server/main.go#L103)

Reply:
Updated and use the rotation of certificate file as the main indicator.

Reply:
Please make sure describe the main file in document? This is very critical design assumption and the cert rotator should treat it as a contract.

Reply:
Updated the design document.